gadjid/src/graph_operations/reachability.rs

```diff
@@ -0,0 +1,722 @@
+            .iter()
+            .filter(|u| !t.contains(*u))
+            .for_each(|u| {
+                next.push((Edge::Undirected, *u));
+            });
+        truth_dag
+            .children_of(v)
+            .iter()
+            .filter(|c| !t.contains(*c))
+            .for_each(|c| {
+                next.push((Edge::Incoming, *c));
+            });
+        next
+    };
+
+    while let Some((arrived_by, node, walkstatus)) = to_visit_stack.pop() {
+        visited.insert((arrived_by, node, walkstatus));
+
+        match walkstatus {
+            Alg6WalkStatus::POSS_DESC_NAM => {
+                not_amenable.insert(node);
+                poss_desc.insert(node);
+            }
+            Alg6WalkStatus::POSS_DESC_AM => {
+                poss_desc.insert(node);
+            }
+            Alg6WalkStatus::DESC => {
+                poss_desc.insert(node);
+                desc.insert(node);
+            }
+            _ => (),
+        }
+
+        for (move_on_by, w) in get_next_steps(node) {
+            let next = match walkstatus {
+                Alg6WalkStatus::Init => match move_on_by {
+                    Edge::Incoming => Some((move_on_by, w, Alg6WalkStatus::DESC)),
+                    Edge::Undirected => Some((move_on_by, w, Alg6WalkStatus::POSS_DESC_NAM)),
+                    _ => None,
+                },
+                Alg6WalkStatus::POSS_DESC_AM | Alg6WalkStatus::POSS_DESC_NAM => match move_on_by {
+                    Edge::Incoming | Edge::Undirected => Some((move_on_by, w, walkstatus)),
+                    _ => None,
+                },
+                Alg6WalkStatus::DESC => match move_on_by {
+                    Edge::Incoming => Some((move_on_by, w, Alg6WalkStatus::DESC)),
+                    Edge::Undirected => Some((move_on_by, w, Alg6WalkStatus::POSS_DESC_AM)),
+                    _ => None,
+                },
+            };
+
+            if let Some(next) = next {
+                if !visited.contains(&next) {
+                    to_visit_stack.push(next);
+                }
+            }
+        }
+    }
+
+    (desc, poss_desc, not_amenable)
+}
+
+/// Validate Z as adjustment set relative to (T, Y) for a given set T of treatment
+/// nodes and all possible Y in G.
+///
+/// Follows Algorithm 7 in https://doi.org/10.48550/arXiv.2402.08616
+///
+/// Returns tuple of:<br>
+/// - Set NVA (Not Validly Adjusted) of nodes Y \notin T in G such that Z is not a valid adjustment set for (T, Y) in G.
+pub fn get_invalid_unblocked(
+    truth_dag: &PDAG,
+    t: &[usize],
+    z: FxHashSet<usize>,
+) -> FxHashSet<usize> {
+    #[allow(non_camel_case_types)]
+    #[derive(Debug, PartialEq, Eq, Hash, Clone, Copy)]
+    enum Alg7WalkStatus {
+        /// Possible Descendant / Partially Directed, and Open Walk
+        PD_OPEN,
+        /// Possible Descendant / Partially Directed, and Blocked Walk
+        PD_BLOCK,
+        /// Non-Causal walk that has not been blocked
+        NON_CAUSAL_OPEN,
```

Comment:
Why is this more consistent? For this algorithm, I think it's useful to highlight that it's a non causal walk that has not been blocked, because blocked non-causal walks are not walked further, which is not directly clear otherwise.

Reply:
This is true for the other NON_CAUSAL walks, too – but we can rename all NON_CAUSAL to NON_CAUSAL_OPEN :-)